news-explorer/src/components/App/App.js

```diff
@@ -92,34 +122,118 @@ function App() {
 		handleScreenResize();
 	}, []);
 
-	function handleDisplayCards() {
+	//recent search resault
+	React.useEffect(() => {
+		if (!allCards) {
+			setCardsData([]);
+			return;
+		}
 		let cardsToRender = [];
-		for (let i = 0; i < count; i++) {
-			if (testData[i]) {
-				cardsToRender.push(testData[i]);
+		let card;
+
+		for (let i = 0; i <= count; i++) {
+			if (allCards[i]) {
+				card = {...allCards[i], id: uuid()};
+				cardsToRender.push(card);
 			}
 		}
-		setCount(count + 3);
 		setCardsData(cardsToRender);
+	}, [allCards, count]);
+
+	function handleDisplayCards() {
+		setCount(count + 3);
 	}
 
-	useEffect(() => {
-		setAllCards(testData);
-		if (cardsData === []) return;
-		function renderCards() {
-			let cardsToRender = [];
-			for (let i = 0; i < 3; i++) {
-				if (!testData[i]) return;
-				cardsToRender.push(testData[i]);
-			}
-			setCardsData(cardsToRender);
-		}
-		renderCards();
-	}, []);
+	function handleSearchSubmit(keyword) {
+		SetIsNewsSearchOpen(true);
+		setIsLoading(true);
+		setCardKeyword(keyword);
+		NewsApi.getArticles(keyword)
+			.then((res) => {
+				if (res.articles.length !== 0) {
+					setAllCards(res.articles);
+
+					localStorage.setItem('cards', JSON.stringify(res.articles));
+				} else {
+					localStorage.removeItem('cards');
+				}
+				setIsLoading(false);
+				SetIsNewsSearchOpen(true);
+			})
+			.catch((err) => {
+				console.log(err);
+			});
+	}
 
 	function handleHamburgerClick() {
 		setIsHamburgerMenuOpen(!isHambuergerMenuOpen);
 	}
+	///get saved articles when user is logged
+	useEffect(() => {
+		if (isLoggedIn) {
+			mainApi
+				.getSavedArticles()
+				.then((res) => {
+					setSaveCardsData(res);
+				})
+				.catch((err) => {
+					console.log(err);
+				});
+		}
+	}, [isLoggedIn]);
+
+	function handleSaveClick(card) {
+		mainApi.markCard(card, cardKeyWord).then((res) => {
+			setSaveCardsData([res, ...savedCardsData]);
+		});
```

Comment:
You need to `catch` possible errors at the very end of any server request.

**Please, correct it everywhere**